tensorpipe/common/cuda_host_allocator.cc

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#include <tensorpipe/common/cuda_host_allocator.h>
+
+#include <cuda_runtime.h>
+
+#include <tensorpipe/common/cuda.h>
+
+namespace tensorpipe {
+
+CudaHostAllocator::CudaHostAllocator(size_t numChunks, size_t chunkSize)
+    : numChunks_(numChunks),
+      chunkSize_(chunkSize),
+      data_(allocPinnedBuffer(numChunks * chunkSize)),
+      chunkAvailable_(numChunks, true) {
+  std::unique_lock<std::mutex> lock(mutex_);
+  thread_ = std::thread([this, lock{std::move(lock)}]() mutable {
+    setThreadName("TP_CUDA_host_allocator_loop");
+    allocLoop(std::move(lock));
+  });
+}
+
+CudaHostAllocator::~CudaHostAllocator() {
+  join();
+}
+
+void CudaHostAllocator::alloc(size_t size, TAllocCallback callback) {
+  std::unique_lock<std::mutex> lock(mutex_);
+  if (closed_) {
+    callback(TP_CREATE_ERROR(CudaHostAllocatorClosedError), THostPtr());
+    return;
+  }
+  TP_DCHECK(size <= chunkSize_);
+  pendingAllocations_.push_back(std::move(callback));
+  cv_.notify_all();
+}
+
+size_t CudaHostAllocator::getChunkLength() const {
+  return chunkSize_;
+}
+
+void CudaHostAllocator::close() {
+  std::unique_lock<std::mutex> lock(mutex_);
+  if (closed_) {
+    return;
+  }
+  closed_ = true;
+  cv_.notify_all();
+}
+
+void CudaHostAllocator::join() {
+  close();
+  if (!joined_.exchange(true)) {
+    thread_.join();
+  }
+}
+
+void CudaHostAllocator::allocLoop(std::unique_lock<std::mutex> lock) {
+  for (;;) {
+    if (closed_ && pendingAllocations_.empty()) {
+      break;
+    } else {
+      cv_.wait(lock);
+    }
+
+    std::deque<TAllocCallback> allocations;
+    std::swap(allocations, pendingAllocations_);
+    lock.unlock();
+    processAllocations(allocations);
+    lock.lock();
+    pendingAllocations_.insert(
+        pendingAllocations_.begin(), allocations.begin(), allocations.end());
+  }
+}
+
+void CudaHostAllocator::processAllocations(
+    std::deque<TAllocCallback>& allocations) {
+  for (size_t curChunk = 0; curChunk < numChunks_; ++curChunk) {
```

Comment:
Is this the most efficient way of going about this? As it stands, we'd risk going over all chunks (a 1000 of them, according to the default values) if the one available chunk is the last one.

We don't really need to know the state of each chunk, all we need to know is which chunks are available. What if we kept a `std::deque<int>` that stored the _indices_ of the available chunks? In `hostPtrDeleter` we would push a new index to the back of that deque, whereas in this method we'd pop from the front until non-empty.

A deque might add some overhead, as it causes allocs and frees when pushing and popping. As in our case the deque size is bounded and known in advance (i.e., up to the total number of chunks), we could reimplement it more efficiently, e.g., as a vector-backed ringbuffer (i.e., push becomes `freeList_[(head++) % numChunks_] = value`, pop becomes `freeList_[(tail++) % numChunks_] if tail < head else -1`).